analyses/upstream-analysis/util/function-process-Seurat.R

```diff
@@ -0,0 +1,322 @@
+  
+  cat("Get cell cycle scores and then scale data", "\n")
+  # Get cell cycle scores (S and G2M) for each cell, Genome specifies which gene set to use for cell cycle genes
+  seurat_obj <- Genome_Specific_Cell_Cycle_Scoring(seurat_obj, Genome)
+  
+  cat("Scale data", "\n")
+  # Indicates whether or not to regress for cell cycle and, if so, which method to use and scale data
+  if (Regress_Cell_Cycle == "NO") {
+    print("We will NOT regress out for cell cycle and scale data!")
+    seurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj))
+    } else if (Regress_Cell_Cycle == "YES") {    
+      print("We will regress out for cell cycle and scale data!")
+      seurat_obj <- ScaleData(seurat_obj, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(seurat_obj))
+      } else if (Regress_Cell_Cycle == "DIFF") {
+        print("We will regress out for cell cycle difference and scale data!")
+        seurat_obj$CC.Difference <- seurat_obj$S.Score - seurat_obj$G2M.Score
+        seurat_obj <- ScaleData(seurat_obj, vars.to.regress = "CC.Difference", features = rownames(seurat_obj))
+  } 
+  
+  cat("Run PCA", "\n") # Get PCs for dimensionality reduction
+  # Either use the prespecified PCA.Feature.List or use Variable Features found before
+  if (missing(PCA_Feature_List)) {
+    seurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj), assay = assay, npcs = num_pcs, verbose = FALSE)
+  } else {
+      seurat_obj <- RunPCA(seurat_obj, features = PCA_Feature_List, assay = assay, npcs = num_pcs, verbose = FALSE)
+  }
+  
+  cat("Run UMAP", "\n")
+  set.seed(1234) # Make code reproducible
+  
+  # Create a dataframe of all of the possible combinations of number of PCs to use for UMAP, and the number of neighbors
+  num_dim_vect <- c(num_dim)
+  num_neighbors_vect <- c(num_neighbors)
+  possibilities <- expand.grid(num_dim_vect, num_neighbors_vect)
+  
+  # For each of these combinations, let's calculate UMAP
+  for(k in 1:nrow(possibilities)) {
+    num_dim <- possibilities[k, 1]
+    num_neighbors <- possibilities[k, 2]
+    seurat_obj <- run_dr(data = seurat_obj, dr_method = "umap", reduction = paste0("pca"),
+                         num_dim_use = num_dim, assay = "RNA", num_neighbors = num_neighbors,
+                         prefix = glue::glue("ndim{num_dim}nn{num_neighbors}"))}
+  
+  # Generate metadata
+  reduction_names <- c(paste0("umap", "ndim", possibilities[,1], "nn", possibilities[,2]), paste0("pca")) # Export the reductions to Seurat
+  metadata <- as_data_frame_seurat(seurat_obj, reduction = reduction_names, metadata = TRUE)
+  
+  # Save output files
+  write_tsv(metadata, file = paste0(results_dir, "/", "metadata", ".tsv")) # Save metadata
+  
+  cat("Plot the plots", "\n")
+  # Identify the 10 most highly variable genes
+  print(top10 <- head(VariableFeatures(seurat_obj), 10))
+  
+  # plot variable features with and without labels
```

Comment:
Super minor, but if the code below will not be used in the module, maybe consider removing just for extra readability

Reply:
It might be useful later.